### PR TITLE
First implementation of the authentication chain with LDAP

### PR DESCRIPTION
**Pull request by albertobucesGdtl**
The authentication chain configuration is found in WebSecurityController. Ldap requires a password encoder in its configuration, it has been set to sha256 (which is marked as deprecated because of its poor security), in case the ldap server used has a different encryption, the encoder would have to be changed.

Includes class LdapUserAuthoritiesPopulator to check the user name in database and get its authorizations when LDAP authentication is active.

The properties for configuring authentication with ldap are in app/src/main/resources/config/application-dev.yml and app/src/main/resources/config/application-prod.yml.

### CODE REVIEW

**Review by fjlopez — CHANGES_REQUESTED** [review]
Please review the comments. Make sure that the tests in `common` and `deploy/heroku-dev-full` pass in local. 

Use an approach similar to https://spring.io/guides/gs/authenticating-ldap/ to validate the behavior of the LDAP support without depending on a Guadaltel LDAP server.

File: common/build.gradle [hunk 1]
```diff
@@ -18,6 +18,9 @@ dependencies {
   implementation 'org.springframework.boot:spring-boot-starter-validation'
   implementation 'org.springframework.boot:spring-boot-starter-security'
   implementation 'org.springframework.boot:spring-boot-starter-actuator'
+  implementation 'org.springframework.ldap:spring-ldap-core'
+  implementation 'org.springframework.security:spring-security-ldap'
+  implementation 'com.unboundid:unboundid-ldapsdk'
```
Comment (fjlopez) [inline common/build.gradle:23]: unboundid-ldapsdk must be only available in tests

File: common/src/main/java/org/sitmun/authentication/controller/AuthenticationController.java [hunk 2]
```diff
@@ -67,4 +67,4 @@ public ResponseEntity<AuthenticationResponse> authenticateUser(@Valid @RequestBo
       return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
     }
   }
-}
+}
```
Comment (fjlopez) [inline common/src/main/java/org/sitmun/authentication/controller/AuthenticationController.java:70]: Add a blank line at the end to be identical

File: common/src/main/java/org/sitmun/infrastructure/security/storage/UserDetailsPasswordStorage.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.sitmun.infrastructure.security.core.userdetails.UserDetailsServiceImplementation;
+import org.springframework.core.annotation.Order;
+import org.springframework.security.config.annotation.authentication.builders.AuthenticationManagerBuilder;
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Component;
+
+@Component
+@Order(2)
```
Comment (fjlopez) [inline common/src/main/java/org/sitmun/infrastructure/security/storage/UserDetailsPasswordStorage.java:11]: Remove (2). Then the default value applies: Ordered.LOWEST_PRECEDENCE. This component must be always the last.

File: deploy/heroku-dev-full/src/test/java/org/sitmun/authentication/controller/AuthenticationControllerTest.java [hunk 4]
```diff
@@ -45,4 +46,32 @@ void loginFailure() throws Exception {
...
+	    login.setUsername("user12");
+	    login.setPassword("user12");
+
+	    mvc.perform(post("/api/authenticate")
+	        .contentType(MediaType.APPLICATION_JSON)
+	        .content(TestUtils.asJsonString(login)))
+	      .andExpect(status().isOk())
+	      .andExpect(jsonPath("$.id_token").exists());
+  }
 }
```
Comment (fjlopez) [inline deploy/heroku-dev-full/src/test/java/org/sitmun/authentication/controller/AuthenticationControllerTest.java:77]: An additional cases must be tested: if the user exists in LDAP but not in the SITMUN database, the login must fail.

**Review by fjlopez — CHANGES_REQUESTED** [review]
Change requests focus on the need of an additional test and  the documentation of the two tests

**Comment by albertobucesGdtl** [comment]
The requested changes have been made.

**Comment by fjlopez** [comment]
The code seems ok but we are waiting some technical specs from Diputaciò Lleida which are required for a final confirmation.

**Comment by albertobucesGdtl** [comment]
Updated LDAP tests

**Comment by fjlopez** [comment]
Some tests were internally broken due to bad configuration options. Redundant files have been removed. 

**Required**

- Tests must be updated to support the LDIF configuration of Diputació de Lleida https://github.com/albertobucesGdtl/sitmun-backend-core/pull/1 


**Actions expected**

- @albertobucesGdtl, do an additional PR with the LDAP tests based on the LDIF configuration of **Diputació de Lleida** rather than on the Spring Framework example.

**Pull request merged**